src/org/openjdk/asmtools/common/outputs/DirOutput.java

```diff
@@ -0,0 +1,98 @@
+package org.openjdk.asmtools.common.outputs;
+
+import org.openjdk.asmtools.common.Environment;
+
+import java.io.BufferedOutputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.nio.file.FileSystems;
+import java.util.Optional;
+
+public class DirOutput extends NamedToolOutput {
+
+    private final File dir;
+    private File outfile;
+    private FileOutputStream fos;
+    private PrintWriter pw;
+
+    public DirOutput(File dir) {
+        this.dir = dir;
+    }
+
+    @Override
+    public String toString() {
+        return super.toString() + " to " + dir;
+    }
+
+    @Override
+    public void printlns(String line) {
+        pw.println(line);
+    }
+
+    @Override
+    public void prints(String line) {
+        pw.print(line);
+    }
+
+    @Override
+    public void prints(char line) {
+        pw.print(line);
+    }
+
+    @Override
+    public void startClass(String fqn, Optional<String> fileExtension, Environment environment) throws IOException {
+        super.startClass(fqn, fileExtension, environment);
+        final String fileSeparator = FileSystems.getDefault().getSeparator();
+        if (dir == null) {
+            int startOfName = fqn.lastIndexOf(fileSeparator);
+            if (startOfName != -1) {
+                fqn = fqn.substring(startOfName + 1);
+            }
+            outfile = new File(fqn + fileExtension.orElseGet(() -> ""));
+        } else {
+            environment.traceln("writing -d " + dir.getPath());
+            if (!fileSeparator.equals("/")) {
```

Comment:
Minor implementation detail: the implementation of `replace` is a no-op if the two arguments are single character strings and equal.  So, there is no need for the external guard in the `if` statement.